### ISSUE TITLE
Unable to delete Reference Data with multiple Reference Data Types

### ISSUE THREAD

**Issue by Schwierig** [akeneo-labs/CustomEntityBundle#134]
I'm unable to delete any reference data since the 2.0 update which added the ProductLinkChecker. The problem likely lies in line 47 in the same class, since it loads all attribute codes and Akeneo automatically validates the code against each of them. This causes Akeneo to throw an InvalidPropertyException.

Example `Property "gen_mat_multi" expects a valid code. No reference data "material" with code "1027" has been found, "1027" given`

I'm using the current CustomEntityBundle version and the 2.1.0 Akeneo version

**Comment by jmleroux** [comment]
Hello @Schwierig ,
Thank you for raising this issue. I don't reproduce the issue so far. Could you please describe your `gen_mat_multi` attribute? It is possible this is a pre-patch issue where you deleted reference data linked to products. In your example, the property is linked to a non-existing ref data.

**Comment by Schwierig** [comment]
So I'm creating two different reference data types and two different reference datas. After that I'm attaching both to a product. If I try to delete one reference data now it will throw the error described above.
This should not be happening post-patch.

**Comment by jmleroux** [comment]
> This should not be happening post-patch.
:+1: 

That's exactly the same kind of tests I made. I will re-check

**Comment by Schwierig** [comment]
Ah I forgot a step. I just tested on a new instance. You have to delete the product first to reproduce the error.

**Comment by jmleroux** [comment]
I'm sorry, but I really cannot reproduce.
I can just suggest you to use Xdebug to see what really happens.

**Comment by Schwierig** [comment]
Hmm that's strange.

I already did that. The Product Link Checker is loading all reference data attribute codes (which isn't really necessary, since I'm only deleting one option of one type). Then it loops through each type and validates against the product builder. Akeneo then throws an error in the ReferenceDataFilter Line 174. Which makes sense, since it validates the code against all reference data types.

This is my stack trace
![screen shot 2018-01-05 at 13 50 57](https://user-images.githubusercontent.com/2803839/34610034-8d05abc6-f21f-11e7-8735-aeeac64d1966.png)

**Comment by jmleroux** [comment]
> which isn't really necessary, since I'm only deleting one option of one type

Alas, it is necessary, because we cannot know to how many attributes is linked this ref data.

**Comment by Schwierig** [comment]
But shouldn't you just select the attributes of the deleted reference data type?

**Comment by jmleroux** [comment]
How would you select them?

**Comment by jmleroux** [comment]
> But shouldn't you just select the attributes of the deleted reference data type?

How would you select them ?

**Comment by Schwierig** [comment]
That's the problem with how attribute properties are stored. The only way to kind of get to the "reference_data_name" would be by using a string search (aka LIKE). But that's not even close to acceptable. At the moment I got really no good idea on how to fix this issue. Tried many things yesterday, didn't really come to a solution.

**Comment by jmleroux** [comment]
Thanks for the stack trace, this will be very helpful

**Comment by jmleroux** [comment]
`$entity->getCode()`? 

https://github.com/akeneo-labs/CustomEntityBundle/blob/master/Checker/ProductLinkChecker.php#L51

This would be weird

**Comment by Schwierig** [comment]
> From where does come the null value?

Not sure which null value you are talking about

**Comment by jmleroux** [comment]
Sorry, my bad. I have another issue running in parallel. And it seems my brain is not good at multi-threading. (It's an old 19s brain. Multithreading is just emulated :) )

What are the arguments values [here](https://github.com/akeneo-labs/CustomEntityBundle/blob/2.0/Checker/ProductLinkChecker.php#L51) when it crashes?

**Comment by Schwierig** [comment]
`$attributeCode = gen_mat_multi` -> Reference Data Type = Material
`$entity->getCode()` -> Reference Data Type = Generic Color

This is the grid of the Generic Color Reference Data Type
![screen shot 2018-01-05 at 14 38 17](https://user-images.githubusercontent.com/2803839/34611506-17e646c8-f226-11e7-99c8-ebac6584f5bf.png)

**Comment by jmleroux** [comment]
> $attributeCode = gen_mat_multi -> Reference Data Type = Material
> $entity->getCode() -> Reference Data Type = Generic Color

I don't understand. You should have something like that:

$attributeCode = `gen_color` and $entity->getCode() = `blau`

or 

$attributeCode = `gen_mat_multi` and $entity->getCode() = `one_material_code`

**Comment by Schwierig** [comment]
`$attributeCodes => Array
(
    [0] => gen_mat_multi
    [1] => ref_color
    [2] => ref_brand
)`

The Entity is of the type `generic_color` which is linked in the `ref_color` attribute. And that's where the error is coming from. Since `gen_mat_multi` is the first attribute code and it isn't linked to the `generic_color` type the validation fails.

**Comment by jmleroux** [comment]
OK! I get it!
The (dirty) solution could be to catch the exception. I'll try to find something cleaner.

**Comment by Schwierig** [comment]
Nice, thanks for all the support :)

**Comment by jmleroux** [comment]
I reproduced and tested, but if you had time to test the PR, you will be more than welcome ;)

**Comment by Schwierig** [comment]
Works for me 👍 

Greetings from Torben :P

**Comment by jmleroux** [comment]
> Greetings from Torben :P

:beers:

**Comment by jmleroux** [comment]
Patch available: https://github.com/akeneo-labs/CustomEntityBundle/releases/tag/v2.0.8
